File: P4-PK/SymbolTable/SymbolTable.java
import java.util.HashMap;

public class SymbolTable {
    public SymbolTable(SymbolTable parentTable) {
        if (parentTable != null) {
            for (Symbol symbol : parentTable.retrieveAllSymbols().values()) {
                st.put(symbol.name, new Symbol(symbol.type, symbol.name, symbol.value, false));
            }
        }
    }

    public Symbol getSymbol(String name) throws GetException {
        try {
            Symbol symbol = st.get(name);
            if (symbol == null) {
                throw new GetException(name);
            }
            return symbol;
        } catch (GetException ge) {
            throw ge;
        }
    }

    public void putSymbol(String type, String name, String value, Boolean local) throws PutException {
        try {
            Symbol symbol = st.get(name);
            if (symbol != null && symbol.local) {
                throw new PutException(name);
            }
            st.put(name, new Symbol(type, name, value, local));
        }
        catch (PutException pe) {
            throw pe;
        }
    }

    public HashMap<String, Symbol> retrieveAllSymbols() {
        return st;
    }

    private HashMap<String, Symbol> st = new HashMap<String, Symbol>();
}

File: Simpleton/SymbolTable/SymbolTable.cs
using System.Collections.Generic;
using Simpleton.AST;

namespace Simpleton.SymTable
{
    public class SymbolTable
    {
        Stack<Dictionary<string, Symbol>> symbolTables = new Stack<Dictionary<string, Symbol>>();

        public SymbolTable()
        {
            OpenNewScope();
        }

        public void OpenNewScope()
        {
            symbolTables.Push(new Dictionary<string, Symbol>());
        }

        public void CloseNewlyCreatedScope()
        {
            symbolTables.Pop();
        }


        public Symbol getSymbol(string name)
        {
            foreach (Dictionary<string, Symbol> table in symbolTables)
            {
                if (table.ContainsKey(name))
                    return table[name];
            }
            throw new GetException(name);
        }

        public void PutSymbol(string name, Symbol symbol)
        {
            if (symbolTables.Peek().ContainsKey(name))
                throw new PutException(name);

            symbolTables.Peek().Add(name, symbol);
        }

        public void Print()
        {

            foreach (Dictionary<string, Symbol> table in symbolTables)
            {
                foreach (string item in table.Keys)
                {
                    System.Console.WriteLine(table[item].ToString());
                }
                System.Console.WriteLine("");
            }
        }
    }

}